### PR TITLE
Achieving algorithm agility. See #1250

### PR DESCRIPTION
**Pull request by ioggstream**
## This PR

Describes how to achieve algorithm agility:

- add Status to IANA table;
- use Want-Digest to negotiate newer algos
- use multiple digest values while transitioning to new algos (see https://tools.ietf.org/html/rfc7696#section-2.3 but I didn't cite it as I don't think it is worth unless we want to add MANDATORY algorithms)

## Note:

I didn't merge this section with IANA Considerations because of [Keep IANA Considerations for IANA]( https://tools.ietf.org/html/rfc8126#section-1.1)

### CODE REVIEW

**Comment by jricher** [comment]
This only addresses the algorithm used in digesting the message body, not the algorithm used to sign content.

**Comment by LPardue** [comment]
No problem @jricher :) This is one of the joys of a shared repo. Labels help but aren't a complete solution.

**Comment by ioggstream** [comment]
> Sorry, ignore that -- I thought this was a comment on the signatures draft, my mistake.

@jricher still you are still in time to  provide some feedback on Digest :D we hope to release it soon!

@LPardue PTAL :P

File: draft-ietf-httpbis-digest-headers.md [hunk 1]
```diff
@@ -271,7 +271,9 @@ representation metadata, payload transformations and HTTP methods on Digest is
 provided in {{examples-unsolicited}} and {{examples-solicited}}.
 
 A `Digest` field MAY contain multiple representation-data-digest values. This
-could be useful for responses expected to reside in caches shared by users with
+could be useful
+to transition from weak algorithms or
+for responses expected to reside in caches shared by users with
 different browsers, for example.
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:277]: If we're going to expand this we can probably write this better. The first thing that spring into my head is something like

A `Digest` field MAY contain multiple representation-data-digest values. For example, a server may provide representation-data-digest values using different algorithms, allowing it to support a population of clients with different evolving capabilities.

Comment (LPardue) [reply draft-ietf-httpbis-digest-headers.md:277]: and optionally tack on

"This is particularly useful in support of transitioning away from weaker algorithms should the need arise; see {{algorithm-agiity}}"

File: draft-ietf-httpbis-digest-headers.md [hunk 2]
```diff
@@ -1059,7 +1061,16 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+Algorithm agility is achieved establishing the IANA Digest Algorithm Values registry,
+and adding it the "Status" field making it possible to deprecate and obsolete insecure algorithms.
+
+Transition from weak algorithms can be achieved
+negotiating the most suitable digest-algorithm using `Want-Digest` (see {{want-digest}})
+or sending multiple representation-data-digest, thus allowing the receiver to select the best
+possible digest-algorithm.
+Even in this case, it is good practice to avoid sending non-necessary
+representation-data-digest values because the receiver could either ignore them (see {{digest}})
+or uselessly consume resources to evaluate all of them.
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1073]: Now that there is text, I can see the bones of what you're trying to say. But I think we can rewrite this to be a bit clearer. Here's strawperson text:

"The security properties of digest algorithms are not fixed. It is in an endpoint's best interest to support and use the most secure algorithms that are at its disposal. However, there is little that this specification can do to mandate behaviour. To help endpoints make better algorithm choices a new "Status" field has been added to the IANA Digest Algorithm Values registry (see Section xyz), which contains the most-recent appraisal of the algorithm. Algorithms that have been found insecure or otherwise undesirable can be marked as "deprecated" or "obsolete". Algorithms with no known problems can be marked as "Standard", while new algorithms can be marked "Experimental".

An endpoint might have a preference for algorithms, such as preferring "standard" algorithms over "deprecated" ones. Algorithm agility is supported by negotiation of digest-algorithm using ``Want-Digest` (see {{want-digest}})` or by sending multiple representation-data-digest values for the receiver to choose. Endpoints are advised that sending multiple values consumes resources, which may be wasted if the receiver ignores them (see {{digest}})."

**Comment by ioggstream** [comment]
@LPardue I tried to integrate your text, which improves significantly the description. Thanks!

If we want to shorten the text, we could replace some parts referencing RFC7696 and pointing the reader there (explanation of algorithm agility, transition from weaker algorithms, ...)

WDYT?

File: draft-ietf-httpbis-digest-headers.md [hunk 3]
```diff
@@ -270,9 +270,14 @@ The relationship between `Content-Location` (see Section 7.8 of
...
-could be useful for responses expected to reside in caches shared by users with
-different browsers, for example.
+A `Digest` field MAY contain multiple representation-data-digest values.
+For example, a server may provide representation-data-digest values using different algorithms,
+allowing it to support a population of clients with different evolving capabilities;
+this is particularly useful in support of transitioning away
+from weaker algorithms should the need arise (see {{algorithm-agility}}).
+Another use case for sending multiple checksums is
+for responses expected to reside in caches shared by users with
+different browsers.
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:280]: This sentence seems to me to just repeat the notion of evolving capability already introduced in the paragraph.

I suspect this text, originally in RFC 3230, was attempting to address transparent caching proxies. I'm not convinced we need to keep it in that form given how many HTTP deployments work. Maybe @mnot has an opinion.

File: draft-ietf-httpbis-digest-headers.md [hunk 4]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
 
 ## Algorithm Agility
 
-...
+The security properties of digest-algorithms are not fixed.
+Algorithm agility is achieved establishing the IANA Digest Algorithm Values
+registry (see {{iana-digest-algorithm-registry}}) and Section 2.1 of {{?RFC7696}}).
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1069]: How about,

> Algorithm Agility (see {{?RFC7696}}) is achieved by providing implementations flexibility in their choice of digest-algorithm from the IANA Digest Algorithm Values registy in {{iana-digest-algorithm-registry}}.

File: draft-ietf-httpbis-digest-headers.md [hunk 5]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+
+To help endpoints make secure algorithm choices,
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1071]: Let's not say secure, because algorithms will inevitably be broken. How about 

```suggestion
To help endpoints understand weaker algorithms from stonger ones,
```

File: draft-ietf-httpbis-digest-headers.md [hunk 6]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+this document adds to the IANA Digest Algorithm Values registry
+a new "Status" field containing the most-recent appraisal of the digest-algorithm according
+to the algorithm used to compute the checksum based on the type and status of the primary document
+in which the checksum algorithm is defined.
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1075]: The remember of the paragraph is hard to parse and I don't think it is strictly true. I'd like to think that we can deprecate algorithms without the need of coordination with the original defining document. The reason some algorithm becomes deprecated or obsolete could because it is a dead. I think it would be simpler to just remove everything after "... appraisal of the digest-algorithm."

File: draft-ietf-httpbis-digest-headers.md [hunk 7]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+
+The entries associated to a checksum algorithm which becomes insecure or otherwise undesirable
+can be marked as "deprecated",
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1078]:
```suggestion
are marked as "deprecated",
```

let's use the active voice.

File: draft-ietf-httpbis-digest-headers.md [hunk 8]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+while the ones associated to a standardized algorithm without known problems can be marked
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1079]:
```suggestion
while the ones associated to a standardized algorithm without known problems are marked
```

File: draft-ietf-httpbis-digest-headers.md [hunk 9]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+as "standard".
+
+An endpoint might have a preference for algorithms,
+such as preferring "standard" algorithms over "deprecated" ones.
+Transition from weak algorithms is supported
+by negotiation of digest-algorithm using `Want-Digest` (see {{want-digest}})
+or by sending multiple representation-data-digest values for the receiver to choose.
```
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1086]:
```suggestion
or by sending multiple representation-data-digest values from which the receiver chooses.
```

**Review by LPardue — CHANGES_REQUESTED** [review]
Ahh I'd overlooked RFC 7696 sorry. I think what's here is almost right, and I've made some minor suggestions.

File: draft-ietf-httpbis-digest-headers.md [hunk 10]
```diff
@@ -1059,11 +1064,32 @@ in conjuction with the encrypted content-coding {{?RFC8188}}.
...
+Endpoints are advised that sending multiple values consumes resources,
+which may be wasted if the receiver ignores them (see {{digest}})."
```
Comment (martinthomson) [inline draft-ietf-httpbis-digest-headers.md:1088]:
```suggestion
which may be wasted if the receiver ignores them (see {{digest}}).
```

File: draft-ietf-httpbis-digest-headers.md (hunk 4 above)
Comment (ioggstream) [inline draft-ietf-httpbis-digest-headers.md:1069]:
```suggestion
Algorithm Agility (see {{?RFC7696}}) is achieved by providing implementations flexibility in their choice of
digest-algorithm from the IANA Digest Algorithm Values registy in {{iana-digest-algorithm-registry}}.
```

Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md]: applied!

File: draft-ietf-httpbis-digest-headers.md (hunk 1 above)
Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md]: Done

File: draft-ietf-httpbis-digest-headers.md (hunk 6 above)
Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:1075]: In our spec  https://github.com/httpwg/http-extensions/blame/master/draft-ietf-httpbis-digest-headers.md#L1079:1082
we reused this part of RFC3864 

      Specify "standard", "experimental", "informational", "historic",
      "obsoleted", or some other appropriate value according to the type
      and status of the primary document in which it is defined.  For
      non-IETF specifications, those formally approved by other
      standards bodies should be labelled as "standard"; others may be
      "informational" or "deprecated" depending on the reason for
      registration.

which is more relaxed than the present. I will relax our statement and ask to reviewers for feedback so that we can land this PR.

File: draft-ietf-httpbis-digest-headers.md (hunk 3 above)
Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:280]: @LPardue Are you proposing this?

```suggestion
```

File: draft-ietf-httpbis-digest-headers.md (hunk 6 above)
Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:1075]:
```suggestion
this document adds to the IANA Digest Algorithm Values registry
a new "Status" field containing the most-recent appraisal of the digest-algorithm.
```

File: draft-ietf-httpbis-digest-headers.md (hunk 3 above)
Comment (LPardue) [reply draft-ietf-httpbis-digest-headers.md:280]: Yes.

**Comment by ioggstream** [comment]
Reorganized the algorithm status here https://github.com/httpwg/http-extensions/pull/1284/commits/d389ea9b0e25de45f4f27885d511dc91f4221631# @LPardue

**Comment by LPardue** [comment]
looks good. Let's land that sentence deletion and merge this.
